main.cu

```diff
@@ -0,0 +1,491 @@
+        ml_params.num_elements = 1;
+        ml_params.elements     = &elem;
+
+        ucs_status_t st;
+        do {
+            st = ucp_device_mem_list_create(eps[p], &ml_params, &mem_lists[p]);
+            if (st == UCS_ERR_NOT_CONNECTED) {
+                ucp_worker_progress(worker);
+            }
+        } while (st == UCS_ERR_NOT_CONNECTED);
+        UCP_CHECK(st, "ucp_device_mem_list_create(per-peer)");
+        element_index[p] = 0; // single element
+    }
+
+    // Build PUT operations: for each peer, compute local and remote offsets
+    std::vector<put_op_t> ops; ops.reserve(world_size);
+    for (int p = 0; p < world_size; ++p) {
+        if (p == rank) continue;
+        put_op_t op;
+        op.list_handle_index = p;
+        op.element_index     = element_index[p];
+        op.address        = (const void*)((uintptr_t)send_buf + senddispls[p]);
+        size_t remote_off = all_recvdispls[p * world_size + rank];
+        op.remote_address = peer_bases[p] + remote_off;
+        op.length         = sendcounts[p];
+        ops.push_back(op);
+    }
+
+    // Upload ops to device
+    put_op_t *d_ops = nullptr;
+    CUDA_CHECK(cudaMalloc(&d_ops, ops.size() * sizeof(put_op_t)));
+    CUDA_CHECK(cudaMemcpy(d_ops, ops.data(), ops.size() * sizeof(put_op_t), cudaMemcpyHostToDevice));
+
+    // Prepare kernel params
+    kernel_params_t kparams = {};
+    const unsigned threads_per_block = 128;
+    const unsigned num_ops = static_cast<unsigned>(ops.size());
+    const unsigned num_blocks = (num_ops + threads_per_block - 1) / threads_per_block;
+    kparams.num_threads  = threads_per_block;
+    kparams.num_blocks   = num_blocks;
+    kparams.level        = UCS_DEVICE_LEVEL_THREAD;
+    kparams.with_request = false;
+    // Upload mem list handle array to device
+    ucp_device_mem_list_handle_h *d_mem_lists = nullptr;
+    CUDA_CHECK(cudaMalloc(&d_mem_lists, sizeof(ucp_device_mem_list_handle_h) * world_size));
+    CUDA_CHECK(cudaMemcpy(d_mem_lists, mem_lists.data(), sizeof(ucp_device_mem_list_handle_h) * world_size, cudaMemcpyHostToDevice));
+    kparams.mem_lists  = d_mem_lists;
+    kparams.num_lists  = world_size;
+
+    // Launch kernel
+    ucs_status_t *d_status = nullptr; ucs_status_t h_status = UCS_OK;
+    CUDA_CHECK(cudaMalloc(&d_status, sizeof(*d_status)));
+    CUDA_CHECK(cudaMemcpy(d_status, &h_status, sizeof(h_status), cudaMemcpyHostToDevice));
+
+    do_alltoallv_kernel<UCS_DEVICE_LEVEL_THREAD><<<kparams.num_blocks, kparams.num_threads>>>(kparams, d_ops, num_ops, d_status);
+    CUDA_CHECK(cudaGetLastError());
+    CUDA_CHECK(cudaDeviceSynchronize());
+    CUDA_CHECK(cudaMemcpy(&h_status, d_status, sizeof(h_status), cudaMemcpyDeviceToHost));
+    if (h_status != UCS_OK) {
+        fprintf(stderr, "Rank %d kernel failed: %d\n", rank, h_status);
+        MPI_Abort(MPI_COMM_WORLD, h_status);
+    }
+    // TODO: Is call to flush necessary?
+    UCP_CHECK(ucp_worker_flush(worker), "ucp_worker_flush");
```

Comment:
I am not sure calling flush is necessary.
However, I might need to call `ucp_device_progress_req` inside the kernel, but this seems to cause a hang...